[Darren Teo] iP

darrentde opened the pull request:
# ChadBot
I am ChadBot and my goal is to be the **biggest bot in town**

```Java
if (i.equals("ChadBot")) {
    System.out.println("Hello ChadBot 🤠");
}
```

## One bot to rule them all
> So good that it is still in development

Features
- Track task
- Save task
- Delete task

Future Plans
1. Become cognitive
2. Auto-pilot coding
3. ~~Enslave humanity~~

Improvements 
- [ ] Unknown commands causing exceptions
- [ ] More unit test
- [ ] CI/CD pipeline
- [ ] Better JavaDocs comments



Learn more about [chat bots](https://en.wikipedia.org/wiki/Chatbot)

Commit:
Add Gradle support

darrentde committed:
Completed level-1

darrentde committed:
Completed level-2

darrentde committed:
Completed level-3

darrentde committed:
Completed level-4

darrentde committed:
Complete Automated Text UI Testing

darrentde committed:
Refactor while loop if else conditions

darrentde committed:
Implement exceptions for level-5

darrentde committed:
Add delete task functionality

darrentde committed:
Complete level-7

Move file processing functionalities to FileHelper

darrentde committed:
Complete level-8

Create datetime in both deadline and event
Add date method to check event/datetime happening at a particular date

Format of datetime for checking must be in d/MM/yyyy

darrentde committed:
Merge branch 'branch-Level-7'

darrentde committed:
Fix import conflicts in chad for merge between level-8 and master

darrentde committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

darrentde committed:
Gradle set up

darrentde committed:
Rename filehelper to storage

darrentde committed:
Refactor to OOP format

Invalid commands do not seem to be caught by exception handler for some reason

darrentde committed:
Add Javadocs to class and methods

darrentde committed:
Create packages for java files

darrentde committed:
Create unit test for Task and Todo

Need to structure test directory either using folders or package

darrentde committed:
Add javadocs for test files

darrentde committed:
Add Checkstyle

darrentde committed:
Merge branch 'branch-A-Checkstyle'

darrentde committed:
Code standardized with checkstyle

darrentde committed:
Implement find functionality to find task based on keyword

darrentde committed:
Merge branch 'branch-A-JavaDoc'

darrentde committed:
Merge branch 'branch-Level-9'

darrentde committed:
Merge branch 'branch-A-CodingStandard'

jovitaanderson commented on src/main/java/chad/Parser.java:
```diff
@@ -0,0 +1,78 @@
+        String keyword = tempArr[0];
+
+        try {
+            switch (keyword) {
```
Good use of switch instead of if/else to prevent deep nesting 🙂

jovitaanderson commented on src/main/java/chad/task/Task.java:
```diff
@@ -0,0 +1,49 @@
+        this.isDone = false;
+    }
+
+    @Override
```
Maybe you could also include javadocs for  the toString() methods to let others know what is printed for the user?

jovitaanderson commented on src/main/java/chad/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+    public static void addEventTask(ArrayList<Task> tasks, String userInput) throws ChadException, IOException {
+        String outputText = "Got it. I've added this task:\n";
+        String[] temp = userInput.replaceFirst("event", "").strip().split("/at");
+        String taskDescription = temp[0].strip();
```
Maybe you could consider combining line 101 and line 102 into one ? Since String[] temp is not used elsewhere 🤔

jovitaanderson commented on src/main/java/chad/Ui.java:
```diff
@@ -0,0 +1,138 @@
+        String keyword = userInput.replaceFirst("find", "").trim();
+        String baseText = "Here are the matching tasks in your list:\n";
+        String output = baseText;
+        int count = 1;
```
Perhaps you could consider changing the variable name to index instead of count, since the output would be printing the index of the task🤔

jovitaanderson reviewed (commented):
Looks good 👍🏼 Majority of your code follows the coding standard, good job 🙂 
LGTM

JordanChua commented on src/main/java/chad/Utility.java:
```diff
@@ -0,0 +1,16 @@
+package chad;
```
Perhaps the method of printText can be abstracted to the UI class instead as a public static method to be used in TaskList as well

JordanChua commented on src/main/java/chad/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+     * @throws IOException Thrown when file cannot be opened in IOException
+     */
+    public static void addDeadlineTask(ArrayList<Task> tasks, String userInput) throws ChadException, IOException {
+        String outputText = "Got it. I've added this task:\n";
```
Maybe this outputText can be defined as a constant String instead since its used when adding an event as well, or abstracted to the UI class

JordanChua commented on src/main/java/chad/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+        tasks.add(newTask);
+
+        String strIsDone = newTask.getDone() ? "1" : "0";
+        Storage.writeToFile("D | " + strIsDone + " | " + newTask.getDescription()
```
Perhaps this can be abstracted to a method writeToFile() for both Events and Deadline since the code is reused

JordanChua commented on src/main/java/chad/Ui.java:
```diff
@@ -0,0 +1,138 @@
+        Utility.printText(outputText);
+    }
+
+    public static void searchTaskByKeyword(ArrayList<Task> tasks, String userInput) {
```
I was wondering if there is a particular reason for implementing the method here instead of within the TaskList class

JordanChua commented on src/main/java/chad/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+     * @throws ChadException Thrown if description or date time is invalid
+     * @throws IOException Thrown when file cannot be opened in IOException
+     */
+    public static void addDeadlineTask(ArrayList<Task> tasks, String userInput) throws ChadException, IOException {
```
Maybe the IOException is not needed since it is will not be thrown in the body of the method

JordanChua commented on src/main/java/chad/TaskList.java:
```diff
@@ -0,0 +1,124 @@
+
+        String[] temp = userInput.replaceFirst("deadline", "").strip().split("/by");
+        String taskDescription = temp[0].strip();
+        String dateTime = temp[1].strip();
```
Perhaps the scenario where the date is entered but in a wrong format can be considered as well since it would be meaningful to tell the user that the date format input is not the correct one

JordanChua reviewed (commented):
Overall the code is really easy to read, although there were some parts of code that were reused and could be abstracted away. Some meaningful errors and exceptions can also be considered :)

alvintfl commented on src/main/java/chad/Chad.java:
```diff
@@ -0,0 +1,33 @@
+     * @param args Reads input
+     * @throws IOException Thrown when helper file cannot be open
+     * @throws ChadException Thrown when invalid error occurs
+     */
```
Perhaps you can include an empty line between the description and parameter section?
Maybe you can also include punctuation behind each parameter description?
I noticed the same issues in several other places too.

alvintfl commented on src/main/java/chad/task/Todo.java:
```diff
@@ -0,0 +1,14 @@
+ * chad.task.Todo task that is child of task
+ */
+public class Todo extends Task {
+    public Todo(String description) {
+        super(description);
+    }
```
Perhaps you can include a Javadoc comment for this constructor? So that users do not need to read the code to understand if the constructor does anything else.

alvintfl commented on src/main/java/chad/Ui.java:
```diff
@@ -0,0 +1,138 @@
+     * @param taskID index of task
+     * @throws IOException throws error if file can't be open in chad.Storage
+     */
+    public static void markTask(ArrayList<Task> tasks, int taskID) throws IOException {
```
Perhaps ```taskID``` can be written as ```taskId``` instead?

alvintfl reviewed (commented):
Overall, the coding standard is good except for some formatting issues with the Javadocs.
Keep up the good work! 👍

darrentde committed:
Work on GUI

darrentde committed:
Finish tutorial for JavaFX

darrentde committed:
Chad bot is able to parse terminal input

darrentde committed:
Fix text parsing

darrentde committed:
Merge branch-level-10

darrentde committed:
Fix exception errors
Able to add random inputs now

darrentde committed:
Add assertion

darrentde committed:
Merge branch 'branch-A-Assertions'

darrentde committed:
Fix code quality

darrentde committed:
Merge branch 'branch-A-CodeQuality'

darrentde committed:
Add archive functionality
Chad is now able to archive tasks in task list
Archived task are cleared from task list

darrentde committed:
Add javadocs to archie functions

darrentde committed:
Merge pull request #2 from darrentde/branch-C-Archive

Branch c archive

darrentde committed:
Changed image box

darrentde committed:
Merge branch 'branch-A-BetterGui'

darrentde committed:
Fix jar issues

darrentde committed:
Add more to UG

darrentde committed:
Rename 1st attempt

darrentde committed:
Rename to Ui
